src/main/java/org/opensearch/knn/index/codec/BasePerFieldKnnVectorsFormat.java

```diff
@@ -129,7 +131,21 @@ public KnnVectorsFormat getKnnVectorsFormatForField(final String field) {
     }
 
     private NativeEngines990KnnVectorsFormat nativeEngineVectorsFormat() {
-        return new NativeEngines990KnnVectorsFormat(new Lucene99FlatVectorsFormat(FlatVectorScorerUtil.getLucene99FlatVectorsScorer()));
+        int buildVectorDatastructureThreshold = getBuildVectorDatastructureThresholdSetting(mapperService.get());
```

Comment:
Need mapper Service to extract those values. If we want to move inside NativeEngine990VectorWriter, then, we need to pass either MapperService or Index Settings. Do you see an option to get this threshold without passing any other parameter from this class?

Reply:
I see,

IMO base class should not be responsible for extracting parameters required for vector formats. I suggest you move this logic by one layer in the `NativeEngine990VectorFormat` and then it can have logic to extract parameters it needs to pass on to reader and writer. If the format needs a dependency on mapper service then it should be fine, we should be injecting the dependency.

Reply:
> IMO base class should not be responsible for extracting parameters required for various vector formats. I suggest you move this logic by one layer in the NativeEngine990VectorFormat and then it can have logic to extract parameters it needs to pass on to reader and writer. If the format needs a dependency on mapper service then it should be fine, we should be injecting the dependency.

Before we take a decision where this logic should be placed(base or codec class), lets first check if the setting is updated then it is getting honored. @VijayanB my question to you would be out of these format classes Base or NativeEngines990KnnVectorsFormat are their instances created for every segment?

Reply:
@navneet1v Yes, new instance is being created every time for new segment creation ( including flush and merge )

Reply:
@shatejas i believe in our offline discussion you don't have strong opinion against passing variable instead of mapper service. After discussion this with @jmazanec15 and @navneet1v , we are thinking about not adding dependency to entire MapperService to Codec, instead, will pass this as parameter to begin with. I will add comment that, if we are going to pass more values to constructor, we should extract those params as POJO and pass one object instead of increasing constructor parameters. Based on discussion, this is aligned with how lucene accepts params for Codec too.  What do you think about this solution?

Reply:
I am not onboard with adding a POJO for the constructor, Its unnecessary to create an intermediate object when you can pass in a dependency, That also doesn't solve the problem of Base class having to manage parameters that are needed by each format that we add moving forward, as it will always have to extract and pass it on to the pojo. What were the reasons discussed as to not pass a dependency (MapperService) when the format needs one?

I am okay to start with one parameter being passed around but I am not onboard with passing a POJO if we need more parameters. We shouldn't be afraid of injecting a dependency when a class needs one especially when there are not blockers in injecting it

Reply:
I think major concern with passing MapperService into format is that formats can be loaded via SPI which requires a no-arg constructor. So, if a codec is loaded in this fashion, it will be impossible to inject the MapperService dependency. I cannot remember exactly when SPI is used - I believe its read only. 

For this reason, though, I dont like adding heavy and complex OpenSearch dependencies into the codec

Reply:
I thought though, we were going to pass in a supplier - I thought that this wouldnt be created for each new segment.

Reply:
@jmazanec15 
So the format instances gets created for every segment, reading and writing. These instances are then part of SegmentReader object. BTW I tested the write path by putting a debugger and I can it creates format 1 for each segment. But yes once the segment is created that instance is now only used for getting the reader.

Now on the part of passing the mapper service. I am with @jmazanec15 lets not pass complex object that cannot be created during SPI loads aka no-args constructor. But in this case our nativeEngineReader currently doesn't have a dependency on the MapperService or the other attributes. So, right now it doesn't matter. But it would be good if we can avoid passing these complex objects to Writer so that we are not blocking ourselves for future, because reader will never have the ability to have these complex objects.

Reply:
@jmazanec15 Given that it is called every time when new segment is created, do you still see need to pass supplier instead of actual value?

Reply:
> But it would be good if we can avoid passing these complex objects to Writer so that we are not blocking ourselves for future

So the idea is to have MapperService in NativeFormat and then it can be used to extract what is needed for writer and skip it for reader. So the writer doesn't have the whole dependency at the same time the base doesn't need to have the knowledge of what each format needs to extract. The reader and writers still don't have complex objects

Reply:
> I think major concern with passing MapperService into format is that formats can be loaded via SPI which requires a no-arg constructor. So, if a codec is loaded in this fashion, it will be impossible to inject the MapperService dependency

@jmazanec15 we can use optional mapperservice here since knnvectorsformat instance is created during runtime. So for SPI load it will have null/empty optional but during runtime creation it can pass in the dependency.

Reply:
@jmazanec15 @VijayanB @navneet1v Lets close this out as I don't think I have issues with the approach currently. we can cross the bridge of if we need the dependency when we have to

Reply:
Ack